common/environment/environment.cpp

```diff
@@ -1589,53 +1587,45 @@ IConstMachineInfo * CLocalEnvironment::getMachine(const char * name) const
     return (CConstMachineInfo *) cached;
 }
 
-IConstMachineInfo * CLocalEnvironment::getMachineByAddress(const char * machineIp) const
+IConstMachineInfo * CLocalEnvironment::getMachineByAddress(const char * hostOrIP) const
 {
-    if (!machineIp)
+    if (isEmptyString(hostOrIP))
         return nullptr;
     buildMachineCache();
-    Owned<IPropertyTreeIterator> iter;
-    StringBuffer xpath;
-    xpath.appendf("Hardware/Computer[@netAddress=\"%s\"]", machineIp);
+
     synchronized procedure(safeCache);
-    IConstEnvBase *cached = getCache(xpath.str());
-    if (!cached)
+
+    VStringBuffer xpath("Hardware/Computer[@netAddress=\"%s\"]", hostOrIP);
+    IConstEnvBase *cached = getCache(hostOrIP);
+    if (cached)
+        return (CConstMachineInfo *) cached;
+
+    IPropertyTree *d = p->queryPropTree(xpath); // exact match
+    if (!d && !isIPAddress(xpath)) // if not found and not an IP, resolve and match against resolved entries
     {
-        IPropertyTree *d = p->queryPropTree(xpath.str());
-        if (!d)
+        IpAddress ip(hostOrIP);
+        Owned<IPropertyTreeIterator> iter = p->getElements("Hardware/Computer");
+        ForEach(*iter)
         {
-            // I suspect not in the original spirit of this but look for resolved IP
-            Owned<IPropertyTreeIterator> iter = p->getElements("Hardware/Computer");
-            IpAddress ip;
-            ip.ipset(machineIp);
-            ForEach(*iter)
+            IPropertyTree &computer = iter->query();
+            IpAddress computerIP;
+            const char *computerNetAddress = computer.queryProp("@netAddress");
+            if (!isEmptyString(computerNetAddress))
             {
-                IPropertyTree &computer = iter->query();
-                IpAddress ip2;
-                const char *ips = computer.queryProp("@netAddress");
-                if (ips&&*ips)
+                // NB: could 1st check if computerNetAddress isIPAddress() and not bother resolving here if it is.
+                computerIP.ipset(computerNetAddress);
+                if (ip.ipequals(computerIP))
                 {
-                    ip2.ipset(ips);
-                    if (ip.ipequals(ip2))
-                    {
-                        d = &computer;
-                        break;
-                    }
+                    d = &computer;
+                    break;
                 }
             }
         }
-        if (!d)
-            return nullptr;
-        StringBuffer xpath1;
-        xpath1.appendf("Hardware/Computer[@name=\"%s\"]", d->queryProp("@name"));
-        cached = getCache(xpath1.str());
-        if (!cached)
-        {
-            cached = new CConstMachineInfo((CLocalEnvironment *) this, d);
-            setCache(xpath1.str(), cached);
-            setCache(xpath.str(), cached);
-        }
     }
+    if (!d)
+        return nullptr;
+    cached = new CConstMachineInfo((CLocalEnvironment *) this, d);
+    setCache(xpath.str(), cached);
```

Comment:
Why not add @name to cache as before ?

Reply:
I should have made a comment in the PR to remember why.
I think it's because it's already in the cache.. built by buildMachineCache initially (and whenever dirty)
So adding by name here was pointless before since already there.

Whereas adding by @netAddress still useful as could be different from what it finds, i.e. could be myhost or myhost.domain, or IP..